bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java

```diff
@@ -0,0 +1,282 @@
+import org.eclipse.launchbar.core.ILaunchBarManager;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.SelectionAdapter;
+import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Combo;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Group;
+import org.eclipse.swt.widgets.Label;
+import org.json.simple.JSONArray;
+
+import com.espressif.idf.core.IDFEnvironmentVariables;
+import com.espressif.idf.core.build.IDFLaunchConstants;
+import com.espressif.idf.core.logging.Logger;
+import com.espressif.idf.core.util.EspConfigParser;
+import com.espressif.idf.core.util.IDFUtil;
+import com.espressif.idf.core.util.StringUtil;
+import com.espressif.idf.launch.serial.internal.SerialFlashLaunchConfigDelegate;
+import com.espressif.idf.launch.serial.util.EspFlashCommandGenerator;
+
+public class CMakeMainTab2 extends GenericMainTab {
+	private static final String EMPTY_CONFIG_OPTIONS = "-s ${openocd_path}/share/openocd/scripts"; //$NON-NLS-1$
+	private Button flashOverJtagButton;
+	private Combo fFlashVoltage;
+	private Combo fTarget;
+	private Map<String, JSONArray> boardConfigsMap;
+	private Combo fTargetName;
+	private boolean isFlashOverJtag;
+	private String defaultArguments;
+	private String argumentsForJtagFlash;
+	private boolean isJtagFlashAvailable;
+
+	@Override
+	public void createControl(Composite parent) {
+		isJtagFlashAvailable = checkIfJtagIsAvailable();
+		super.createControl(parent);
+		createJtagFlashButton(parent);
+	}
+
+	private void createJtagFlashButton(Composite parent) {
+		if (!isJtagFlashAvailable) {
+			return;
+		}
+		flashOverJtagButton = new Button(parent, SWT.CHECK);
+		flashOverJtagButton.setText(Messages.CMakeMainTab2_JtagComboLbl);
+		flashOverJtagButton.addSelectionListener(new SelectionListener() {
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				isFlashOverJtag = !isFlashOverJtag;
+				if (!isFlashOverJtag) {
+					argumentField.setText(defaultArguments);
+				} else {
+					argumentField.setText(argumentsForJtagFlash);
+				}
+				switchUI();
+			}
+
+			@Override
+			public void widgetDefaultSelected(SelectionEvent e) {
+			}
+		});
+	}
+
+	private void switchUI() {
+		if (!isJtagFlashAvailable) {
+			return;
+		}
+		locationField.setEnabled(!isFlashOverJtag);
+		workDirectoryField.setEnabled(!isFlashOverJtag);
+		fileLocationButton.setEnabled(!isFlashOverJtag);
+		workspaceLocationButton.setEnabled(!isFlashOverJtag);
+		variablesLocationButton.setEnabled(!isFlashOverJtag);
+		fileWorkingDirectoryButton.setEnabled(!isFlashOverJtag);
+		workspaceWorkingDirectoryButton.setEnabled(!isFlashOverJtag);
+		variablesWorkingDirectoryButton.setEnabled(!isFlashOverJtag);
+
+		fFlashVoltage.setEnabled(isFlashOverJtag);
+		fTargetName.setEnabled(isFlashOverJtag);
+		fTarget.setEnabled(isFlashOverJtag);
```

Comment:
If none of these are needed for JTAG flashing, probably we can move JTAG flashing to the top and show the required UI based on the selection.

Reply:
@kolipakakondal 
good point. UI is changed